webapp/graphite/storage.py

```diff
@@ -37,33 +41,58 @@ def __init__(self, finders=None, hosts=None):
   def find(self, pattern, startTime=None, endTime=None, local=False, headers=None):
     query = FindQuery(pattern, startTime, endTime, local)
 
+    for match in self.find_all(query, headers):
+      yield match
+
+
+  def find_all(self, query, headers=None):
+    start = time.time()
+    result_queue = Queue.Queue()
+    jobs = []
+
     # Start remote searches
     if not query.local:
       random.shuffle(self.remote_stores)
-      remote_requests = [ r.find(query, headers) for r in self.remote_stores if r.available ]
+      jobs.extend([
+        (store.find, query, headers)
+        for store in self.remote_stores if store.available
+      ])
 
-    matching_nodes = set()
-
-    # Search locally
+    # Start local searches
     for finder in self.finders:
-      for node in finder.find_nodes(query):
-        #log.info("find() :: local :: %s" % node)
-        matching_nodes.add(node)
+      jobs.append((finder.find_nodes, query))
 
-    # Gather remote search results
-    if not query.local:
-      for request in remote_requests:
-        for node in request.get_results():
-          #log.info("find() :: remote :: %s from %s" % (node,request.store.host))
-          matching_nodes.add(node)
+    if settings.USE_WORKER_POOL:
+      return_result = lambda x: result_queue.put(x)
+      for job in jobs:
+        get_pool().apply_async(func=job[0], args=job[1:], callback=return_result)
+    else:
+      for job in jobs:
+        result_queue.put(job[0](*job[1:]))
 
     # Group matching nodes by their path
-    nodes_by_path = {}
-    for node in matching_nodes:
-      if node.path not in nodes_by_path:
-        nodes_by_path[node.path] = []
+    nodes_by_path = defaultdict(list)
+
+    deadline = start + settings.REMOTE_FIND_TIMEOUT
+    result_cnt = 0
+
+    while result_cnt < len(jobs):
+      if time.time() > deadline:
+        log.info("Timed out in find_all after %fs" % (settings.REMOTE_FIND_TIMEOUT))
+        break
+
+      try:
+        nodes = result_queue.get(True, 0.01)
```

Comment:
That's to make that `while` loop because on each iteration it's checking again if the `deadline` has been reached or not. 
Or is your question more about why exactly `0.01` and not any other number? I think it just needed to be some small number to re-check the `deadline` often.